File: Zampa-iOS/Zampa/Core/LocalizationManager.swift
import SwiftUI
import FirebaseFirestore

final class LocalizationManager: ObservableObject {
    static let shared = LocalizationManager()

    @Published var currentLanguage: String {
        didSet {
            UserDefaults.standard.set(currentLanguage, forKey: "appLanguage")
            loadStrings()
        }
    }

    private var strings: [String: String] = [:]

    static let supportedLanguages: [(code: String, nativeName: String)] = [
        ("auto", "Automático"),
        ("es", "Español"),
        ("ca", "Català"),
        ("eu", "Euskara"),
        ("gl", "Galego"),
        ("en", "English"),
        ("de", "Deutsch"),
        ("fr", "Français"),
        ("it", "Italiano"),
    ]

    private init() {
        self.currentLanguage = UserDefaults.standard.string(forKey: "appLanguage") ?? "auto"
        loadStrings()
    }

    var resolvedLanguage: String {
        if currentLanguage == "auto" {
            let systemLang = Locale.current.language.languageCode?.identifier ?? "es"
            let supported = Self.supportedLanguages.map { $0.code }.filter { $0 != "auto" }
            return supported.contains(systemLang) ? systemLang : "es"
        }
        return currentLanguage
    }

    var resolvedLanguageNativeName: String {
        if currentLanguage == "auto" {
            let resolved = resolvedLanguage
            let systemName = Self.supportedLanguages.first { $0.code == resolved }?.nativeName ?? "Español"
            return "Automático (\(systemName))"
        }
        return Self.supportedLanguages.first { $0.code == currentLanguage }?.nativeName ?? currentLanguage
    }

    private func loadStrings() {
        let lang = resolvedLanguage
        guard let url = Bundle.main.url(forResource: lang, withExtension: "json", subdirectory: nil),
              let data = try? Data(contentsOf: url),
              let dict = try? JSONDecoder().decode([String: String].self, from: data) else {
            // Fallback to Spanish
            if lang != "es",
               let url = Bundle.main.url(forResource: "es", withExtension: "json", subdirectory: nil),
               let data = try? Data(contentsOf: url),
               let dict = try? JSONDecoder().decode([String: String].self, from: data) {
                strings = dict
            }
            return
        }
        strings = dict
    }

    func t(_ key: String) -> String {
        strings[key] ?? key
    }

    func setLanguage(_ code: String, userId: String?) {
        currentLanguage = code
        guard let userId = userId else { return }
        Firestore.firestore().collection("users").document(userId).updateData([
            "languagePreference": code
        ])
    }

    func syncFromFirebase(userId: String) {
        Firestore.firestore().collection("users").document(userId).getDocument { [weak self] snapshot, _ in
            if let lang = snapshot?.data()?["languagePreference"] as? String, !lang.isEmpty {
                DispatchQueue.main.async {
                    self?.currentLanguage = lang
                }
            }
        }
    }
}

File: Zampa-Android/app/src/main/java/com/sozolab/zampa/data/LocalizationManager.kt
package com.sozolab.zampa.data

import android.content.Context
import android.content.res.Configuration
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

data class SupportedLanguage(val code: String, val nativeName: String)

@Singleton
class LocalizationManager @Inject constructor(
    private val context: Context
) {
    companion object {
        val supportedLanguages = listOf(
            SupportedLanguage("auto", "Automático"),
            SupportedLanguage("es", "Español"),
            SupportedLanguage("ca", "Català"),
            SupportedLanguage("eu", "Euskara"),
            SupportedLanguage("gl", "Galego"),
            SupportedLanguage("en", "English"),
            SupportedLanguage("de", "Deutsch"),
            SupportedLanguage("fr", "Français"),
            SupportedLanguage("it", "Italiano"),
        )
        private val supportedCodes = supportedLanguages.map { it.code }.filter { it != "auto" }
    }

    private val prefs = context.getSharedPreferences("zampa_prefs", Context.MODE_PRIVATE)

    private val _currentLanguage = MutableStateFlow(prefs.getString("appLanguage", "auto") ?: "auto")
    val currentLanguage: StateFlow<String> = _currentLanguage

    val resolvedLanguage: String
        get() {
            val lang = _currentLanguage.value
            if (lang == "auto") {
                val systemLang = Locale.getDefault().language
                return if (systemLang in supportedCodes) systemLang else "es"
            }
            return lang
        }

    val resolvedLanguageNativeName: String
        get() {
            val lang = _currentLanguage.value
            if (lang == "auto") {
                val resolved = resolvedLanguage
                val name = supportedLanguages.find { it.code == resolved }?.nativeName ?: "Español"
                return "Automático ($name)"
            }
            return supportedLanguages.find { it.code == lang }?.nativeName ?: lang
        }

    fun createLocalizedContext(): Context {
        val locale = Locale(resolvedLanguage)
        val config = Configuration(context.resources.configuration)
        config.setLocale(locale)
        return context.createConfigurationContext(config)
    }

    fun setLanguage(code: String, userId: String? = null) {
        _currentLanguage.value = code
        prefs.edit().putString("appLanguage", code).apply()
        userId?.let { uid ->
            FirebaseFirestore.getInstance()
                .collection("users")
                .document(uid)
                .update("languagePreference", code)
        }
    }

    fun syncFromFirebase(userId: String) {
        FirebaseFirestore.getInstance()
            .collection("users")
            .document(userId)
            .get()
            .addOnSuccessListener { doc ->
                val lang = doc.getString("languagePreference")
                if (!lang.isNullOrEmpty()) {
                    _currentLanguage.value = lang
                    prefs.edit().putString("appLanguage", lang).apply()
                }
            }
    }
}
